doc/using/organizing.chapter.md

```diff
@@ -0,0 +1,76 @@
+# Organizing packages with callPackage {#chap-organizing}
+
+You can use [`stdenv.mkDerivation`](#sec-using-stdenv) to build packages, but it quickly becomes cumbersome when managing more than a small handful of packages. Nixpkgs provides a number of tools to help build and manage large sets of packages which depend on each other or on the Nixpkgs package set.
+
+## callPackage {#sec-callpackage}
+
+`pkgs.callPackage` is a tool for filling function arguments with `pkgs` attributes. For instance, if you have a file `foo.nix`:
+
+```nix
+{
+  stdenv,
+  fetchurl,
+  zlib,
+}:
+stdenv.mkDerivation {
+  name = "libfoo-1.2.3";
+  src = fetchurl {
+    url = "http://example.org/libfoo-1.2.3.tar.bz2";
+    hash = "sha256-tWxU/LANbQE32my+9AXyt3nCT7NBVfJ45CX757EMT3Q=";
+  };
+  nativeBuildInputs = [zlib];
+}
+```
+
+Then `pkgs.callPackage ./foo.nix { }` is equivalent to `import ./foo.nix { inherit (pkgs) stdenv fetchurl zlib; }`. The second argument to `callPackage` can be used to override arguments or to specify extra arguments that aren't in Nixpkgs.
+
+Nixpkgs itself is built from many packages in different files, combined together with `callPackage` invocations.
+
+## makeScope {#sec-makescope}
+
+When you're building up a package set, you'll typically want your packages to depend on each other, like these two packages:
+
+```nix
+# my-cool-library.nix
+{
+  stdenv,
+}:
+stdenv.mkDerivation {
+  name = "my-cool-library-1.2.3";
+  # ...
+}
+```
+
+```nix
+# my-cool-program.nix
+{
+  stdenv,
+  my-cool-library,
+}:
+stdenv.mkDerivation {
+  name = "my-cool-program-1.2.3";
+  buildInputs = [my-cool-library];
+  # ...
+}
+```
+
+With `pkgs.callPackage`, we can compose a package set like this:
+
+```nix
+# The `rec` keyword is generally discouraged; read on for more.
+rec {
+  my-cool-library = pkgs.callPackage ./my-cool-library.nix { };
+  my-cool-program = pkgs.callPackage ./my-cool-program.nix { inherit my-cool-library; };
+}
+```
+
+This quickly becomes frustrating, so Nixpkgs provides a tool to extend `callPackage` with custom packages, [`lib.mkScope`](#function-library-lib.customisation.makeScope). We can rewrite the previous example with `mkScope` like this, replacing `pkgs.callPackage` with `self.callPackage`:
+
+```nix
+lib.mkScope pkgs.newScope (self: {
```

Comment:
It would also be worthwhile to say something about `pkgs.newScope`.

Reply:
I agree, but I don't know enough about `newScope` to write documentation for it. This type signature makes my head hurt:

```
AttrSet -> ((AttrSet -> a) | Path) -> AttrSet -> a
```

Reply:
Added a little note, although my understanding is incomplete. Let me know if it looks good.